### DISCUSSION TITLE
Don't highlight messages as new in a different window with the same split(s) as the current one

### DISCUSSION THREAD

**Discussion by hemirt** [Chatterino/chatterino2#5412] [Ideas] [answered]
Having two tabs with the same channel split inside would highlight the other window as receiving a new unknown message, despite it coming from the same channel that is already being viewed in the current tab.
![image](https://github.com/Chatterino/chatterino2/assets/1310440/0ad48de3-59c7-4331-89a0-cd0fb0faf71e)
![image](https://github.com/Chatterino/chatterino2/assets/1310440/e886a2c5-e702-4580-ae82-e3685c2b25ee)

**Comment by pajlada** [comment]
This makes sense - I don't see it needing an option either, the behaviour you're talking about should just be the default.

If you have 3 tabs, 2 contain your channel. you're currently in the tab not containing your channel. A new message comes in on your channel.
Both tabs containing your channel should highlight, right?
When you select one of those tabs, should both tabs containing your channels be marked as unread (assuming the message in your channel was the only new one)?

**Reply by hemirt** [reply]
hopefully I understand it correctly, I think reading the message in one tab should mark all others tab containing this message as already read

whilst ofcourse if youre not in a tab with your channel, all tabs containing that channel should highlight

so all tabs should remember which channel made the tab show the unread notification, and if you go into another tab that contains some channels, all other tabs should remove those unread flags, and only keep the unread notification if there are other channels with unread messages

**Reply by hemirt** [reply]
same for mentions or other highlights

**Comment by hemirt** [comment]
Is there a way to know which channels (/splits) does the current tab (NotebookTab / Notebook) contain? I tried looking at how to implement this feature and the best I could figure out is to know who is causing the highlightState change of a NotebookTab (via the invoke slot), so I could add who is causing highlight changes (actors) to a NotebookTab but I could not find a way to figure out which splits are currently displayed, i.e. which highlight changes actors I should ignore -> i.e. 1) not highlighting other tabs due to actors already being viewed in current tab.

I would presume that 2) unhighlighting other tabs on tab change based on actors in the newly being viewed tab, would come out of that easily.

I think then a Notebook, which has its multiple NotebookTab s, would be a good place to do it.

Therefore how I think this could be implemented:
1) https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/splits/SplitContainer.cpp#L220 would add a parameter defining itself (i.e. split name, twitch channel name, split memory address, or just some other pointer)
2) Notebook would know currently being viewed actors and NotebookTab s could query this (via this->notebook_) and ignore highlightState change requests from these actors. https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/helper/NotebookTab.cpp#L361-L382
3) On NotebookTab selection, other tabs would be processed to remove actors from the newly selected tab, and if it was a last actor (i.e. the list of actors causing highlightStateChange is empty), it would unhiglight itself. https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/helper/NotebookTab.cpp#L305-L312


So as I mentioned, is there a way to get the currently being viewed splits/channels?

**Reply by Nerixyz** [reply]
> So as I mentioned, is there a way to get the currently being viewed splits/channels?

Yes, using [`SplitNotebook::getSelectedPage`](https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/Notebook.hpp#L226-L227).

What about message highlights in `/mentions` (https://github.com/Chatterino/chatterino2/issues/1185) and message highlights in general? The main issue I see with unhighlighting `/mentions` if the source channel is selected is that we don't know if the user is in front of their PC. If you leave for a few minutes while viewing a fast chat and get mentioned, `/mentions` should be highlighted (as the message might get lost in chat). But if you're actively reading the chat and get mentioned (i.e. you see the mention), `/mentions` shouldn't be highlighted. We can't really detect that. For new-message highlights, this isn't an issue.

> which highlight changes actors I should ignore

What's a highlight actor? A TwitchChannel? A TwitchChannel + filters? Ideally, these actors can be compared efficiently, since there could be a lot of comparisons (for each "duplicate" actor - 1, for each window, for each split in the selected tab).

**Reply by hemirt** [reply]
> > So as I mentioned, is there a way to get the currently being viewed splits/channels?
> 
> Yes, using [`SplitNotebook::getSelectedPage`](https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/Notebook.hpp#L226-L227).

I see, I only saw a https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/Notebook.hpp#L112 

> What about message highlights in `/mentions` (#1185) and message highlights in general? The main issue I see with unhighlighting `/mentions` if the source channel is selected is that we don't know if the user is in front of their PC. If you leave for a few minutes while viewing a fast chat and get mentioned, `/mentions` should be highlighted (as the message might get lost in chat). But if you're actively reading the chat and get mentioned (i.e. you see the mention), `/mentions` shouldn't be highlighted. We can't really detect that. For new-message highlights, this isn't an issue.

I think /mentions should probably just be left of its own (i.e. not doing anything for it in regards to unhighlighting) - since there's probably a reason you even have this tab open.

> > which highlight changes actors I should ignore
> 
> What's a highlight actor? A TwitchChannel? A TwitchChannel + filters? Ideally, these actors can be compared efficiently, since there could be a lot of comparisons (for each "duplicate" actor - 1, for each window, for each split in the selected tab).

Thats what I was asking for, since it's the split's channelView signal https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/splits/SplitContainer.cpp#L216-L222, then it makes sense for it to be the channelView.channel() (or maybe if hasSourceChannel == true then sourceChannel (I believe these are like popup windows or sth)), comparing pointer shouldn't be a performance issue -> although I see it could be a problem with plenty of them, granted I dont think there would be a lot of comparisons done for it to cause issues, but instead some signal slots could be used, should it be an issue, such signal slots would get updated accordingly on new split / changed channel / deleted split.


> What's a highlight actor? A TwitchChannel? A TwitchChannel + filters?

No idea what filters are, highlight actor in the end, is the one causing the highlightState to change, and as far as I could see, it was only this https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/splits/SplitContainer.cpp#L220 and Notebook::duplicatePage https://github.com/Chatterino/chatterino2/blob/85a7f4a6a91b5fede338f8810975fa63973856d2/src/widgets/Notebook.cpp#L216

**Reply by hemirt** [reply]
I looked at filters and I see the possible issue you eluded to. Those should probably be independent, in my opinion, whilst the direction from filtered channels to unfiltered is fine and probably expected, the other direction from unfiltered to filtered is probably not expected, as again as with mentions, you probably have a reason to have filters.

That is, since if a message appears in filtered channel, it most certainly has appeared at the same time in the underlying channel, therefore from filter to underlying channel it should unhighlight.

But the other way from underlying channel to filter, it should not.

(Ofcourse being completely independent is another possible choice)

**Comment by hemirt** [comment, marked as answer]
https://github.com/Chatterino/chatterino2/pull/5649
